crates/services/upgradable-executor/src/instance.rs

```diff
@@ -0,0 +1,574 @@
+            .map_err(|e| {
+                ExecutorError::Other(format!(
+                    "Failed definition of the `{}` function: {}",
+                    method, e
+                ))
+            })?;
+        Ok(())
+    }
+}
+
+/// This stage adds host functions to get transactions from the transaction source.
+pub struct Source;
+
+impl Instance<Created> {
+    /// Adds host functions for the `source`.
+    pub fn add_source<TxSource>(
+        mut self,
+        source: Option<TxSource>,
+    ) -> ExecutorResult<Instance<Source>>
+    where
+        TxSource: TransactionsSource + Send + Sync + 'static,
+    {
+        let source = source.map(|source| Arc::new(source));
+        let peek_next_txs_size = self.peek_next_txs_size(source.clone());
+        self.add_method("peek_next_txs_size", peek_next_txs_size)?;
+
+        let consume_next_txs = self.consume_next_txs();
+        self.add_method("consume_next_txs", consume_next_txs)?;
+
+        Ok(Instance {
+            store: self.store,
+            linker: self.linker,
+            stage: Source {},
+        })
+    }
+
+    fn peek_next_txs_size<TxSource>(&mut self, source: Option<Arc<TxSource>>) -> Func
+    where
+        TxSource: TransactionsSource + Send + Sync + 'static,
+    {
+        let closure = move |mut caller: Caller<'_, ExecutionState>,
+                            gas_limit: u64|
+              -> anyhow::Result<u32> {
+            let Some(source) = source.clone() else {
+                return Ok(0);
+            };
+
+            let txs: Vec<_> = source
+                .next(gas_limit)
+                .into_iter()
+                .map(|tx| match tx {
+                    MaybeCheckedTransaction::CheckedTransaction(checked) => {
+                        let checked: Checked<Transaction> = checked.into();
+                        let (tx, _) = checked.into();
+                        tx
+                    }
+                    MaybeCheckedTransaction::Transaction(tx) => tx,
+                })
+                .collect();
+
+            let encoded_txs = postcard::to_allocvec(&txs).map_err(|e| {
+                ExecutorError::Other(format!(
+                    "Failed encoding of the transactions for `peek_next_txs_size` function: {}",
+                    e
+                ))
+            })?;
+            let encoded_size = u32::try_from(encoded_txs.len()).map_err(|e| {
+                ExecutorError::Other(format!(
+                    "The encoded transactions are more than `u32::MAX`. We support only wasm32: {}",
+                    e
+                ))
+            })?;
+
+            caller
+                .data_mut()
+                .next_transaction
+                .entry(encoded_size)
+                .or_default()
+                .push(encoded_txs);
+            Ok(encoded_size)
+        };
+
+        Func::wrap(&mut self.store, closure)
+    }
+
+    fn consume_next_txs(&mut self) -> Func {
```

Comment:
nit: this consumes only a single tx
```suggestion
    fn consume_next_tx(&mut self) -> Func {
```